### DISCUSSION TITLE
单测编写有没有可视化的ui辅助

### DISCUSSION THREAD

**Discussion by David-Liu5460** [ant-design/ant-design#51229] [Q&A]
对于复杂的UI组件，如果硬编写在没有可视化ui（dom tree）的情况下写断言是不是相对困难，antd怎么处理的，可以探讨一下吗？
![Uploading image.png…]()

**Comment by petercat-assistant** [comment]
在编写复杂UI组件的单元测试时，确实可能会遇到在没有可视化UI（如DOM树）的情况下编写断言的困难。Ant Design 在这方面通常采用以下几种方法来处理：

1. **使用测试库**：Ant Design 通常使用 React Testing Library 和 Jest 进行单元测试。这些工具提供了一些方法来模拟用户交互，并检查组件的渲染输出。

2. **快照测试**：通过快照测试，可以捕获组件的渲染输出，并在后续测试中与快照进行比较。这种方法可以帮助检测组件输出的意外变化。

3. **可视化测试工具**：虽然 Ant Design 本身可能没有内置的可视化测试工具，但可以结合使用一些第三方工具，如 Storybook 或 Chromatic，这些工具可以帮助在开发过程中可视化组件，并进行交互测试。

4. **调试工具**：使用浏览器的开发者工具可以帮助查看组件的DOM结构和样式，这在编写和调试测试时非常有用。

如果你在使用 Ant Design 时遇到具体的测试问题，欢迎详细描述，我会尽力帮助你解决。

如需进一步帮助，请在评论中描述您的问题，并@petercat-assistant与我开始对话。

**Comment by Wxh16144** [comment]
眼熟你，上周才问过怎么写 unit test，我的建议是先看文档，从别人的 PR 中找到技巧。

> 硬编写在没有可视化ui（dom tree）的情况下写断言

看文档：[queries](https://testing-library.com/docs/queries/about)

针对组件，你可以把 antd 的 website 启动起来(`npm run start`) 然后用浏览器控制台审查元素即可。举个例子

_button.tsx_

```tsx
export function Button({ className, children }) {
  return (
    <button className={['my-btn', className].filter(Boolean).join(' ')}>
      {children}
    </button>
  )
}
```

_button.test.tsx

```tsx
import { render } from "@testing-library/react";
import { Button } from "./button";

describe("Button", () => {
  it("应该支持传入className", () => {
    const { container, getByRole,getAllByText } = render(<Button className="my-class">按钮</Button>);

    expect(container.firstChild).toHaveClass("my-class"); // 断言是否有 my-class 这个 class

    expect(getByRole("button")).toHaveTextContent("按钮"); // 断言是否有按钮这个文本

    expect(getAllByText("按钮")).toHaveClass("my-class"); // 断言是否有 my-class 这个 class
  })
});
```

**Reply by Wxh16144** [reply]
另外提一嘴，你的提问的图片挂了...
<img width="641" alt="image" src="https://github.com/user-attachments/assets/9aad6d62-9996-41fa-bdd3-b58e377e3bef">
